include/kernel.h

```diff
@@ -1417,6 +1421,309 @@ struct k_timer {
+ * no effect in other alarm states.
+ *
+ * If the flag is present the alarm will be rescheduled to the new
+ * deadline.  `K_ALARM_FLAG_ERROR_IF_LATE` applies to the attempt to
+ * reschedule.
+ *
+ * This flag supersedes `K_ALARM_FLAG_IF_SOONER`.
+ */
+#define K_ALARM_FLAG_REPLACE 0x02
+
+/** Override an existing scheduled alarm if the new deadline is closer.
+ *
+ * This flag overrides the behavior of `k_alarm_schedule()` when
+ * invoked with an already-scheduled alarm that has not fired; it has
+ * no effect in other alarm states.
+ *
+ * If the flag is present the alarm will be rescheduled to the new
+ * deadline only if that is sooner than the existing deadline;
+ * otherwise the existing deadline will be left unchanged.
+ *
+ * `K_ALARM_ERROR_IF_LATE` applies if the alarm is rescheduled.
+ *
+ * This flag is superseded by `K_ALARM_FLAG_REPLACE`.
+ */
+#define K_ALARM_FLAG_IF_SOONER 0x04
+
+typedef void (*k_alarm_handler_t)(struct k_alarm *alarm,
+				  void *ud);
+
+typedef void (*k_alarm_cancel_t)(struct k_alarm *alarm,
+				 void *ud);
+
+struct k_alarm {
+	/* links the alarm into the scheduled and ready queues. */
+	sys_dnode_t node;
+
+	/* Low 32-bits of system clock at which the alarm should fire. */
+	u32_t deadline;
+
+	/* runs in ISR context */
+	k_alarm_handler_t handler_fn;
+
+	/* runs in the context of the thread that calls k_alarm_cancel() */
+	k_alarm_cancel_t cancel_fn;
+
+	/* user-specific data passed to handler and cancel functions. */
+	void *user_data;
+
+	/* the state of the alarm. */
+	enum k_alarm_state state;
+};
+
+#define Z_ALARM_INITIALIZER(handler, cancel, ud) (struct k_alarm) \
+	{ \
+	.node = {},\
+	.deadline = 0, \
+	.handler_fn = handler, \
+	.cancel_fn = cancel, \
+	.user_data = ud, \
+	.state = K_ALARM_UNSCHEDULED, \
+	}
+
+#define K_ALARM_DEFINE(name, handler, cancel, user_data)	\
+	struct k_alarm name					\
+	__in_section(_k_alarm, static, name) =			\
+		Z_ALARM_INITIALIZER(handler, cancel, user_data)
+
+/** Ensure the RTC driving the system clock is configured to generate
+ * an interrupt for the next scheduled deadline.
+ *
+ * This is invoked by the alarm infrastructure whenever the head of
+ * the scheduled alarm list changes.  The implementation should be
+ * provided by the platform-specific system timer module.
+ *
+ * @note This is not user API.
+ */
+void z_alarm_update_deadline(void);
+
+/** Determine how long until alarm processing is required.
+ *
+ * @note This is not user API.
+ *
+ * @param deadlinep pointer to a location to store the absolute
+ * deadline of the first scheduled alarm.  The referenced object is
+ * only modified if this function returns a positive value.
+ *
+ * @retval negative if no alarms are ready or scheduled.
+ *
+ * @retval 0 if at least one alarm is in the ready queue.
+ *
+ * @retval positive if there is a scheduled alarm.  *deadlinep will
+ * contain the corresponding deadline, which may be in the past.
+ */
+int k_alarm_next_deadline_(u32_t *deadlinep);
```

Comment:
remove the _ at the end of the name. (This is not used in any other kernel function afaik)